### PR TITLE
Feature 1103 implementing owasp zap wrapper

### LINKED ISSUE #1103: Implement OWASP ZAP Wrapper for PDS
## Situation
For #607 we need a wrapper.

## Wanted
We want a cli tool to interact with the OWASP ZAP using the OWASP ZAP client API. This shall make the configuration and automation of scans easier.

## Solution
Implementing an cli tool which 
- uses the `SecHubWebScanConfiguration` to  configure the OWASP ZAP
- executes scans with the OWASP ZAP
- collects a report for the executed scan
- can be extended to perform more confirguration for future needs

so no manual interaction with the OWASP ZAP is needed.

### PR DESCRIPTION
**Pull request by winzj**
Implements OWASP ZAP wrapper supporting unauthenticated scans and HTTP basic authentication scans.

- closes #1103

### CODE REVIEW

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+				recurse,
+				contextName,
+				subtreeonly);
+		/* @formatter:on */
+        waitForSpiderResults(responseSpider);
+    }
+
+    @Override
+    protected void runAjaxSpider() throws ClientApiException {
+        String inscope = "true";
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java:41]:
```suggestion
        String inScope = "true";
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        String subtreeonly = "true";
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java:42]:
```suggestion
        String subTreeOnly = "true";
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        String contextname = scanConfig.getContextName();
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java:43]:
```suggestion
        String contextName = scanConfig.getContextName();
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        String url = scanConfig.getTargetUrlAsString();
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java:44]:
```suggestion
        String targetUrl = scanConfig.getTargetUrlAsString();
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		/* @formatter:on */
+        waitForAjaxSpiderResults(responseAjaxSpider);
+    }
+
+    @Override
+    protected void runActiveScan() throws ClientApiException {
+        String url = scanConfig.getTargetUrlAsString();
+        String inscopeonly = "true";
+        String recurse = "true";
+        String scanpolicyname = null;
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/UnauthenticatedScan.java:61]:
```suggestion
        String scanPolicyName = null;
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/HTTPBasicAuthScan.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        String port = Integer.toString(scanConfig.getTargetUri().getPort());
+        /* @formatter:off */
+		StringBuilder authmethodconfigparams = new StringBuilder();
+		authmethodconfigparams.append("hostname=").append(urlEncodeUTF8(scanConfig.getTargetUri().getHost()))
+							  .append("&realm=").append(urlEncodeUTF8(realm))
+							  .append("&port=").append(urlEncodeUTF8(port));
+		/* @formatter:on */
+        LOG.info("For scan {}: Setting authentication.", scanConfig.getContextName());
+        String authmethodname = scanConfig.getAuthenticationType().getOwaspZapAuthenticationMethod();
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/HTTPBasicAuthScan.java:46]:
```suggestion
        String authMethodName = scanConfig.getAuthenticationType().getOwaspZapAuthenticationMethod();
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/EnvironmentVariableReader.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.owaspzapwrapper.util;
+
+public class EnvironmentVariableReader {
+
+    /**
+     *
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/EnvironmentVariableReader.java:7]: What is this method doing? Please provide a short description.

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            SSLContext sslContext = SSLContext.getInstance("SSL");
+            sslContext.init(null, trustAllCerts, new SecureRandom());
+            HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
+        } catch (KeyManagementException | NoSuchAlgorithmException e) {
+            LOG.error("An exception occurred while installing the all-trusting trust manager for checking if target URL is reachable: {}",
+                    targetUrl.toExternalForm(), e);
+        }
+    }
+
+    private static HostnameVerifier createAllTrustingHostNameVerifier() {
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java:113]:
```suggestion
    private static HostNameVerifier createAllTrustingHostNameVerifier() {
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        HostnameVerifier allHostsValid = new HostnameVerifier() {
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java:114]:
```suggestion
        HostNameVerifier allHostsValid = new HostnameVerifier() {
```

File: sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/ScanDurationHelperTest.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        /* execute */
+        long ajaxSpiderDuration = scanDurationHelperToTest.computeAjaxSpiderMaxScanDuration(activeScanEnabled, maxScanDuration);
+        long remainingMaxDuration = maxScanDuration - ajaxSpiderDuration;
+        long spiderDuration = scanDurationHelperToTest.computeSpiderMaxScanDuration(activeScanEnabled, ajaxSpiderEnabled, remainingMaxDuration);
+        remainingMaxDuration = remainingMaxDuration - spiderDuration;
+        long passiveScanDuration = scanDurationHelperToTest.computePassiveScanMaxScanDuration(activeScanEnabled, ajaxSpiderEnabled, remainingMaxDuration);
+        long activeScanDuration = remainingMaxDuration - passiveScanDuration;
+
+        /* test */
+        assertTrue(ajaxSpiderDuration == 4000);
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/ScanDurationHelperTest.java:90]: Please use `assertEquals` instead of `assertTrue`.

```suggestion
        assertEquals(ajaxSpiderDuration, 4000);
```

File: sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubIncludeExcludeToOwaspZapURIHelperTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        assertTrue(urls.contains("https://127.0.0.1:8080/sub.*"));
+        assertTrue(urls.contains("https://127.0.0.1:8080/sub1/directory.*"));
+        assertTrue(urls.contains("https://127.0.0.1:8080/sub2/directory.*"));
+        assertTrue(urls.contains("https://127.0.0.1:8080/sub3/directory.*"));
+        assertTrue(urls.contains("https://127.0.0.1:8080/sub4/directory.*"));
+    }
+
+    private List<String> createExampleListOfSites() {
+        List<String> subSites = new ArrayList<>();
+        subSites.add("/sub/");
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubIncludeExcludeToOwaspZapURIHelperTest.java:66]: What happens if I want to include or exclude a single URL? For example:`https://127.0.0.1:8080/xyz/index.html`. Does that work?

File: sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionCheckerTest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    void edge_case_repsonse_code_400_returns_false() {
+        /* test */
+        // edge case is not allowed
+        assertFalse(TargetConnectionChecker.isReponseCodeValid(400));
+    }
+
+    @Test
+    void not_allowed_response_code_return_false() {
+        /* test */
+        assertFalse(TargetConnectionChecker.isReponseCodeValid(404));
```
Comment (Jeeppler) [inline sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionCheckerTest.java:41]: Please use a parameterized test here.

**Review by Jeeppler — CHANGES_REQUESTED** [review]
See comments.

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java (hunk 8 above)
Comment (winzj) [reply sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java:113]: This is an interface `javax.net.ssl.HostnameVerifier` that was just used here.

File: sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubIncludeExcludeToOwaspZapURIHelperTest.java (hunk 11 above)
Comment (winzj) [reply sechub-wrapper-owasp-zap/src/test/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubIncludeExcludeToOwaspZapURIHelperTest.java:66]: Single directories cannot be excluded right now, because the regular expression `.*` is appended to every URI. The URI from your example would be changed to `https://127.0.0.1:8080/xyz/index.html.*`, which works since it is a file. If we want to exclude single directories, I have to think of a different solution.

**Review by Jeeppler — CHANGES_REQUESTED** [review]
Overall it looks good. However, I still found some smaller issues.

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/cli/OwaspZapScanConfigurationFactory.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            zapHost = environmentVariableReader.readAsString(EnvironmentVariableConstants.ZAP_HOST_ENV_VARIABLE_NAME);
+        }
+        if (zapPort <= 0) {
+            zapPort = environmentVariableReader.readAsInt(EnvironmentVariableConstants.ZAP_PORT_ENV_VARIABLE_NAME);
+        }
+        if (zapApiKey == null) {
+            zapApiKey = environmentVariableReader.readAsString(EnvironmentVariableConstants.ZAP_API_KEY_ENV_VARIABLE_NAME);
+        }
+
+        if (zapHost == null || zapPort <= 0 || zapApiKey == null) {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/cli/OwaspZapScanConfigurationFactory.java:90]: I think @Jeeppler is absolutely right here. It is not the same variable which is checked, so it should not be handled in same line

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubWebScanConfigurationHelper.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.owaspzapwrapper.helper;
+
+import com.mercedesbenz.sechub.commons.model.SecHubTimeUnit;
+import com.mercedesbenz.sechub.commons.model.SecHubWebScanConfiguration;
+import com.mercedesbenz.sechub.commons.model.login.WebLoginConfiguration;
+import com.mercedesbenz.sechub.owaspzapwrapper.config.auth.AuthenticationType;
+
+public class SecHubWebScanConfigurationHelper {
+    private static final int DEFAULT_MAX_SCAN_DURATION = 28800000; // 8 hours in milliseconds
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubWebScanConfigurationHelper.java:10]: Please change to
```java

private static final int SECONDS_IN_MS=1000;
private static final int MINUTES_IN_MS=60*SECONDS_IN_MS;
private static final int HOURS_IN_MS=60*MINUTES_IN_MS;

private static final int DEFAULT_MAX_SCAN_DURATION = 8 * HOURS_IN_MS;
```

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubWebScanConfigurationHelper.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        if (!sechubWebConfig.getMaxScanDuration().isPresent()) {
+            return DEFAULT_MAX_SCAN_DURATION;
+        }
+
+        SecHubTimeUnit sechubTimeUnit = sechubWebConfig.getMaxScanDuration().get().getUnit();
+        int maxScanDuration = sechubWebConfig.getMaxScanDuration().get().getDuration();
+
+        switch (sechubTimeUnit) {
+        case DAY:
+            return maxScanDuration * 1000 * 60 * 60 * 24;
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/helper/SecHubWebScanConfigurationHelper.java:37]: Please use constants mentioned before

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+        this.remainingScanTime = scanConfig.getMaxScanDurationinMillis();
+        this.includeExcludeConverter = new SecHubIncludeExcludeToOwaspZapURIHelper();
+    }
+
+    @Override
+    public void scan() {
+        try {
+            scanUnsafe();
+        } catch (ClientApiException e) {
+            LOG.error("For scan {}: An error occured while scanning! Reason: {}", scanConfig.getContextName(), e.getMessage(), e);
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:65]: Ahem. The scan is not working. The exception is logged. okay. but... how is the failure returned to the caller script? 

When a PDS bash script uses the owasp zap wrapper and we have here a client api exception, how shall the bash script be able to recoginze that there was a problem?

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+        boolean timeOut = false;
+
+        while (!isAjaxSpiderStopped(ajaxSpiderStatus) && !timeOut) {
+            waitForNextCheck();
+            ajaxSpiderStatus = ((ApiResponseElement) clientApi.ajaxSpider.status()).getValue();
+            LOG.info("For scan {}: AjaxSpider status {}", scanConfig.getContextName(), ajaxSpiderStatus);
+
+            if (scanConfig.isVerboseOutput()) {
+                String numberOfResults = ((ApiResponseElement) clientApi.ajaxSpider.numberOfResults()).getValue();
+                List<ApiResponse> spiderResults = ((ApiResponseList) clientApi.ajaxSpider.results("0", numberOfResults)).getItems();
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:150]: What is this magic "0" ? If it is a magic constant, we should provide a constant in java as well + dedicated name so it's clear what is meant.

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+            renameReportFileIfFileExtensionIsNotJSON();
+
+            LOG.info("For scan {}: Report can be found at {}", scanConfig.getContextName(), reportFile.toFile().getAbsolutePath());
+
+            // clean up after scan only if report was done correctly
+            // to investigate errors we keep the session otherwise, because a new scan
+            // always starts a new session, if it worked here or not does not matter
+            cleanUp();
+        } catch (ClientApiException e) {
+            LOG.error("For scan {}: Error writing report file", scanConfig.getContextName(), e);
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:265]: Again... back channel to pds caller script... ?

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+
+        Path parent = reportFile.getParent();
+        Path absolutePath = parent.toAbsolutePath();
+
+        return absolutePath.toString();
+    }
+
+    private void renameReportFileIfFileExtensionIsNotJSON() {
+        String reportFile = scanConfig.getReportFile().toAbsolutePath().toFile().getAbsolutePath();
+        if (!reportFile.endsWith(".json")) {
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:297]: ??? Why/When does this happen?

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+            try {
+                Path owaspzapReport = Paths.get(reportFile + ".json");
+                Files.move(owaspzapReport, owaspzapReport.resolveSibling(scanConfig.getReportFile().toAbsolutePath()), StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                LOG.error("For scan {}: Could not rename report file.", scanConfig.getContextName(), e);
+            }
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:303]: back channel? how do sent back to PDS caller script that we had problems

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/AbstractAuthScan.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+        super(clientApi, scanConfig);
+    }
+
+    @Override
+    public void scan() {
+        try {
+            scanUnsafe();
+        } catch (ClientApiException e) {
+            LOG.error("For scan {}: An error occured while scanning! Reason: {}", scanConfig.getContextName(), e.getMessage(), e);
+        }
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/AbstractAuthScan.java:47]: Backchannel...

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/HTTPBasicAuthScan.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+		authMethodConfigParams.append("hostname=").append(urlEncodeUTF8(scanConfig.getTargetUri().getHost()))
+							  .append("&realm=").append(urlEncodeUTF8(realm))
+							  .append("&port=").append(urlEncodeUTF8(port));
+		/* @formatter:on */
+        LOG.info("For scan {}: Setting authentication.", scanConfig.getContextName());
+        String authMethodName = scanConfig.getAuthenticationType().getOwaspZapAuthenticationMethod();
+        clientApi.authentication.setAuthenticationMethod(contextId, authMethodName, authMethodConfigParams.toString());
+
+        String methodName = SessionManagementType.HTTP_AUTH_SESSION_MANAGEMENT.getOwaspZapSessionManagementMethod();
+        clientApi.sessionManagement.setSessionManagementMethod(contextId, methodName, null);
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/auth/HTTPBasicAuthScan.java:50]: Why is the last parameter null? what about a local variable and a short comment why we set this null?

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+            throw new IllegalArgumentException("Target URI: " + targetUri + " could not be transformed into URL!");
+        }
+        if ("https".equals(urlToCheckConnection.getProtocol())) {
+            TrustManager[] trustAllCerts = createTrustAllCertsManager();
+
+            installAllTrustingTrustManager(urlToCheckConnection, trustAllCerts);
+
+            HostnameVerifier allHostsValid = createAllTrustingHostNameVerifier();
+            // Install the all-trusting host verifier
+            HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
```
Comment (de-jcup) [inline sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/util/TargetConnectionChecker.java:58]: Critical: do NOT install the via the static method! Please look into `TrustAllSupport` of sechub.

Reason: Doing this EVERY standard ssl context URL connection is trusted afterwards - by EVERY url call. So when somebody uses the wrapper as a library after the reachable check, every other url fetch where a url connection is done standard SSL context, every URL would be trusted as well!

**Review by de-jcup — CHANGES_REQUESTED** [review]
Good job, but there are some issues. Please look at my comments.

File: sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java (hunk 17 above)
Comment (winzj) [reply sechub-wrapper-owasp-zap/src/main/java/com/mercedesbenz/sechub/owaspzapwrapper/scan/AbstractScan.java:148]: I added a comment to this section of the code. The "0" is the index, because if verbose output is enabled we log the currently available results. The ZAP API uses a String here, but it is index 0 to mark start with the first and log all currently available.

**Review by de-jcup — APPROVED** [review]
Good job :-)
